soccer/planning/tests/BezierPathTest.cpp

```diff
@@ -6,35 +6,98 @@
-                              constraints);
+static void check_bezier_low_curvature(const Planning::BezierPath& path) {
+    // Expected error is O(1/N)
+    constexpr int N = 1000;
+    double ds = 1.0 / static_cast<double>(N);
 
-    const int N = 1000;
+    for (int i = 0; i <= N; i++) {
+        double s = i * ds;
+        double curvature = 0;
+        path.Evaluate(s, nullptr, nullptr, &curvature);
 
-    std::ofstream out("/tmp/bezier.csv");
+        EXPECT_LE(std::abs(curvature), 100);
+    }
+}
 
-    for (int i = 0; i <= N; i++) {
-        double s = i / static_cast<double>(N);
+static void check_bezier_smooth(const Planning::BezierPath& path) {
+    // Expected error decreases with high N
+    constexpr int N = 10000;
+    constexpr double epsilon = 1e-2;
+
+    Point previous_position;
+    Point previous_velocity;
+    path.Evaluate(0, &previous_position, &previous_velocity, nullptr);
+
+    double ds = 1.0 / static_cast<double>(N);
+
+    for (int i = 1; i <= N; i++) {
+        double s = i * ds;
         Point position;
         Point tangent;
         double curvature = 0;
         path.Evaluate(s, &position, &tangent, &curvature);
 
-        out << s << ", " << position.x() << ", " << position.y() << ", "
-            << tangent.x() << ", " << tangent.y() << ", " << curvature
-            << std::endl;
+        EXPECT_LE((0.5 * (previous_velocity + tangent) * ds).distTo(position - previous_position), epsilon);
+
+        double curvature_expected = (tangent.normalized() - previous_velocity.normalized()).mag() / ds / tangent.mag();
+
+        // Make sure that the approximate curvature is consistent with the calculated exact value.
+        EXPECT_NEAR(curvature, std::abs(curvature_expected), epsilon);
 
-        EXPECT_LT(curvature, 100);
+        previous_position = position;
+        previous_velocity = tangent;
     }
-}
\ No newline at end of file
+}
+
+TEST(BezierPath, two_points_path_smooth_and_consistent) {
+    MotionConstraints constraints;
+    std::vector<Point> points{Point{0, 0}, Point{1, 1}};
+    Planning::BezierPath path(std::move(points), Point(1, 0), Point(1, 0),
+                              constraints);
+    check_bezier_smooth(path);
+}
+
+TEST(BezierPath, multiple_points_path_smooth_and_consistent) {
+    MotionConstraints constraints;
+    std::vector<Point> points{Point{0, 0}, Point{1, 1}, Point{2, 0}};
+    Planning::BezierPath path(std::move(points), Point(1, 0), Point(1, 0),
+                              constraints);
+    check_bezier_smooth(path);
+}
+
+// Smoothed paths should have reasonably low curvature everywhere on the
+// path. Sadly, this does not hold with our current system, which uses
+// Bezier curves and places keypoints in a fairly naive manner.
+// All examples with zero-velocity endpoints are broken because of numerical issues (in this case, the second and third control points go on top of the first and fourth, respectively).
```

Comment:
Might need to make pretty

Reply:
Yeah I'm doing that in a separate part, it's kind of annoying to run make pretty on each of them individually.